src/caffeine/caffeine.c

```diff
@@ -53,110 +53,77 @@ void c_sync_all()
   gasnet_barrier_notify(0,GASNET_BARRIERFLAG_ANONYMOUS);
   gasnet_barrier_wait(0,GASNET_BARRIERFLAG_ANONYMOUS);
 }
-void c_co_sum_no_result_image_int32(void* c_loc_a, size_t Nelem, int* stat)
+
+void c_co_reduce_universal(void* c_loc_a, size_t Nelem, int* stat, int* result_image, int gex_dt, size_t sizeof_type, int gex_op)
```

Comment:
Slapping an `inline` keyword on `c_co_reduce_universal` is probably a good idea, although a good enough optimizer might figure that out.

If you want a portable demand for inlining, you can instead use `GASNETT_INLINE(c_co_reduce_universal)` immediately before the function prototype ([docs](https://gasnet.lbl.gov/dist-ex/README-tools))

The motivation is that inlining allows the GASNet call to "see" the manifest constant parameters provided in the callers; I just checked and this doesn't actually make much difference for the current Reduce implementation, but it makes a big difference for some other GASNet communication calls (notably RMA and atomics).

Reply:
@bonachea Sounds good.  I'll try `GASNETT_INLINE`.  Out of curiosity, why the double `T?

Reply:
GASNETT_ is the namespace for the "GASNet Tools" interface, a collection of portability tools for running on different platforms/compilers/etc that are not directly relevant to communication, but provided as a service to our clients.

GASNet tools has its own spec here: https://gasnet.lbl.gov/dist-ex/README-tools

Reply:
@bonachea I tried inserting `GASNETT_INLINE(c_co_reduce_universal)` on the line before `void c_co_reduce_universal(...` and got the compiler warning and errors below.  Because you mention "function prototype," I wonder if the problem is that I don't have a header file. I also thought the issue might be that the file isn't getting preprocessed so I tried adding `--cpp` in the `--c-flag` argument in `build/run-fpm.sh`.  

```
././src/caffeine/caffeine.c:57:1: warning: return type defaults to 'int' [-Wimplicit-int]
   57 | GASNETT_INLINE(c_co_reduce_universal)
      | ^~~~~~~~~~~~~~
././src/caffeine/caffeine.c: In function 'GASNETT_INLINE':
././src/caffeine/caffeine.c:59:1: error: expected '=', ',', ';', 'asm' or '__attribute__' before '{' token
   59 | {
      | ^
././src/caffeine/caffeine.c:74:1: error: expected '=', ',', ';', 'asm' or '__attribute__' before '{' token
   74 | {
      | ^
```
and more errors similar to the above ones.

Reply:
> @bonachea I tried inserting `GASNETT_INLINE(c_co_reduce_universal)` on the line before `void c_co_reduce_universal(...` and got the compiler warning and errors below. Because you mention "function prototype," I wonder if the problem is that I don't have a header file. I also thought the issue might be that the file isn't getting preprocessed so I tried adding `--cpp` in the `--c-flag` argument in `build/run-fpm.sh`.

I'm not sure what might be wrong without seeing the code you actually compiled - are you 100% certain you did not delete the `void` return type? The macro goes _before_ the return type, something like:

```
GASNETT_INLINE(c_co_reduce_universal)
void c_co_reduce_universal(void* c_loc_a, size_t Nelem, int* stat, int result_image, int dt, size_t sizeof_type, int op, gex_Coll_ReduceFn_t* user_op, void* client_data)
{
... function body...
```

It doesn't matter whether you have a header file or not, this should work regardless.

Reply:
Note that you ALSO need to `#include <gasnet_tools.h>` at the top - that's probably why it failed to compile..

Reply:
added in 34e3cbc95de2d7971c3f698a58bfead5e573e51c